### PR TITLE
feat(function): support length function for Array & Array<T>

### LINKED ISSUE #5225: Feature: length function should work with Array & Array<T> columns
**Summary**

Description for this feature.

```
:) create table a (number Array(UInt8) );
Query OK, 0 rows affected (0.03 sec)
Read 0 rows, 0.00 B in 0.005 sec., 0 rows/sec., 0.00 B/sec.

:) insert into a values( [1,2]), ([3,4]);
Query OK, 0 rows affected (0.04 sec)
Read 2 rows, 28.00 B in 0.011 sec., 174.21 rows/sec., 2.38 KiB/sec.

:) select * from a;
+--------+
| number |
+--------+
| [1, 2] |
| [3, 4] |
+--------+
2 rows in set (0.03 sec)
Read 2 rows, 28.00 B in 0.004 sec., 490.61 rows/sec., 6.71 KiB/sec.

:) select length(number) from a;
ERROR 1105 (HY000): Code: 1007, displayText = Expected a string type, but got Array(UInt8) (while in select before projection).
:)

```

### PR DESCRIPTION
**Pull request by fkuner**
I hereby agree to the terms of the CLA available at: https://databend.rs/dev/policies/cla/

## Summary

provide `length` function for Array & Array\<T>

## Changelog

- New Feature
- Build/Testing/CI

## Related Issues

Fixes #5225

### CODE REVIEW

File: common/functions/src/scalars/semi_structureds/array_length.rs [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+        &self,
+        _func_ctx: FunctionContext,
+        columns: &ColumnsWithField,
+        input_rows: usize,
+    ) -> Result<ColumnRef> {
+        let array_column: &ArrayColumn = Series::check_get(columns[0].column())?;
+        let mut res: Vec<u64> = Vec::with_capacity(input_rows);
+
+        for array in array_column.iter() {
+            let len = array.to_owned_scalar().values.len() as u64;
```
Comment (sundy-li) [inline common/functions/src/scalars/semi_structureds/array_length.rs:75]: I think we just need the offset column.

Comment (fkuner) [reply common/functions/src/scalars/semi_structureds/array_length.rs:75]: yes, make sense

Comment (b41sh) [reply common/functions/src/scalars/semi_structureds/array_length.rs:75]: we can use function `fn len(&self) -> usize` of `ArrayColumn`, which is `offsets.len() -1`

Comment (b41sh) [reply common/functions/src/scalars/semi_structureds/array_length.rs:75]: we can use function `pub fn size_at_index(&self, i: usize) -> usize` of `ArrayColumn`

Comment (fkuner) [reply common/functions/src/scalars/semi_structureds/array_length.rs:75]: > we can use function `pub fn size_at_index(&self, i: usize) -> usize` of `ArrayColumn`

yeah, thanks

File: tests/suites/0_stateless/02_function/02_0053_function_semi_structureds_array_length.sql [hunk 2]
```diff
@@ -0,0 +1,7 @@
+DROP TABLE IF EXISTS array_length_test;
+CREATE TABLE array_length_test(number Array(UInt8));
```
[the remark this answers is not in the document]
Comment (fkuner) [reply tests/suites/0_stateless/02_function/02_0053_function_semi_structureds_array_length.sql:2]: make sense

File: common/functions/src/scalars/semi_structureds/length.rs [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+use common_exception::ErrorCode;
+use common_exception::Result;
+
+use crate::scalars::Function;
+use crate::scalars::FunctionContext;
+use crate::scalars::FunctionDescription;
+use crate::scalars::FunctionFeatures;
+
+#[derive(Clone)]
+pub struct VariantArrayLengthFunction {
```
[the remark this answers is not in the document]
Comment (fkuner) [reply common/functions/src/scalars/semi_structureds/length.rs:29]: But I don't support the Variant except for VariantArray. It depends on whether we support it.

clickhouse's [jsonLength](https://clickhouse.com/docs/en/sql-reference/functions/json-functions/)

File: common/functions/src/scalars/semi_structureds/length.rs [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+        &self,
+        _func_ctx: FunctionContext,
+        columns: &ColumnsWithField,
+        input_rows: usize,
+    ) -> Result<ColumnRef> {
+        let array_column: &VariantColumn = Series::check_get(columns[0].column())?;
+        let mut res: Vec<u64> = Vec::with_capacity(input_rows);
+
+        for array in array_column.iter() {
+            if !array.is_array() {
```
[the remark this answers is not in the document]
Comment (sundy-li) [reply common/functions/src/scalars/semi_structureds/length.rs:75]: I think it's better to throw error as now.

```
select length(3333);
```

It's strange to return 1 as the number has 4 length.

[the remark this answers is not in the document]
Comment (b41sh) [reply common/functions/src/scalars/semi_structureds/length.rs:75]: ok

**Review by sundy-li — APPROVED** [review]

**Pull request merged**